src/stacked_borrows.rs

```diff
@@ -406,8 +383,13 @@ impl<'tcx> Stack {
             // Remove everything above the write-compatible items, like a proper stack. This makes sure read-only and unique
             // pointers become invalid on write accesses (ensures F2a, and ensures U2 for write accesses).
             let first_incompatible_idx = self.find_first_write_incompatible(granting_idx);
-            for item in self.borrows.drain(first_incompatible_idx..).rev() {
-                trace!("access: popping item {:?}", item);
+            let items_to_remove = self.len() - first_incompatible_idx;
+            // This mutation of the stack temporarily jumbles some internal state. This guard
+            // repairs that state on drop, and while it exists it only permits operations which
```

Comment:
So, we assume the guard is never leaked?

I guess that's fine, though a closure-based API seems better to me.